samples/webcodecs-echo/js/stream_worker.js

```diff
@@ -0,0 +1,685 @@
+     self.postMessage({severity: 'info', text: 'Decoder Queue report: ' + JSON.stringify(decqueue_stats)});
+     if (stopped) return;
+     stopped = true;
+     this.stopped = true;
+     self.postMessage({severity: 'fatal', text: 'stop() called'});
+     // TODO: There might be a more elegant way of closing a stream, or other
+     // events to listen for.
+     if (encoder.state != "closed") encoder.close();
+     if (decoder.state != "closed") decoder.close();
+     self.postMessage({severity: 'fatal', text: "stop(): frame, encoder and decoder closed"});
+     return;
+   }
+
+   SendStream(self, transport) {
+     return new WritableStream({
+       async start(controller) {
+         // called by constructor
+         // test to see if transport is still usable?
+       },
+       async write(chunk, controller) {
+         try {
+           let stream = await transport.createUnidirectionalStream();
+           let writer = stream.getWriter();
+           await writer.write(chunk);
+           await writer.close();
+         } catch (e) {
+           self.postMessage({severity: 'fatal', text: 'Failiure to write chunk'});
+         }
+       }, 
+       async close(controller) {
+         // close the transport? 
+         await transport.close();
+       }, 
+       async abort(reason) {
+         // called when ws.abort(reason)
+         // close the transport?
+         await transport.close(); 
+       } 
+     });
+   }
+
+   ReceiveStream(self, transport) {
+     return new ReadableStream({
+       async start(controller) {
+       // called by constructor
+         this.streamNumber = 0;
+         this.reader = transport.incomingUnidirectionalStreams.getReader();
+       },
+       async pull(controller) {
+         // called read when controller's queue is empty
+         let stream_reader, number;
+         try {
+           const { value, done } = await this.reader.read();
+           if (done) {
+             controller.close();
+             self.postMessage({severity: 'fatal', text: 'Done accepting unidirectional streams'});
+             return;
+           }
+           number = this.streamNumber++;
+           //self.postMessage({text: 'New incoming stream # ' + number});
+           stream_reader = value.getReader();
+         } catch (e) {
+           self.postMessage({severity: 'fatal', text: `Error in obtaining stream.getReader(), stream # ${number} : ${e.message}`});
+         }
+         controller.enqueue(await get_frame(stream_reader, number));
+       },
+       async cancel(reason){
+         // called when cancel(reason) is called
+         controller.close();
+         self.postMessage({severity: 'fatal', text: `Readable Stream Cancelled: ${reason}`});
+       }
+     });
+   }
+
+   async start()
+   {
+     if (stopped) return;
+     started = true;
+     self.postMessage({text: 'Start method called.'});
+     try { 
+       this.inputStream
```

Comment:
missing `await` needed to handle stream errors properly.

Strategies to run both this `this.ReceiveStream` below in parallel include using IFFEs, or assigning promise from each to `const p1` and `const p2` respectively and doing a `Promise.all([p1, p2])` (with or without individual error handling).

Reply:
Thanks. Will investigate.....